### PR TITLE
Update generic class attribute and type variable usage

### PR DESCRIPTION
**Pull request by pkch**
Clarification based on python/mypy#2878

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: pep-0484.txt [hunk 1]
```diff
@@ -610,6 +611,21 @@ the runtime class of the objects created by instantiating them doesn't
...
+  # (continued from previous example)
+  Node[int].x = 1  # Error
+  Node[int].x      # Error
+  Node.x = 1       # Error
+  Node.x           # Error
+  type(p).x        # Error
+  p.x              # Ok
+  Node[int]().x    # Ok
+
+In addition, ``ClassVar`` cannot be parametrized.
```
Comment (JelleZijlstra) [inline pep-0484.txt]: parameterized

File: pep-0484.txt [hunk 2]
```diff
@@ -610,6 +611,21 @@ the runtime class of the objects created by instantiating them doesn't
 record the distinction. This behavior is called "type erasure"; it is
 common practice in languages with generics (e.g. Java, TypeScript).
 
+Class attribute of a generic class can be looked up only through the instance
```
Comment (DimitrisJim) [inline pep-0484.txt]: "attributes" is probably what you meant here.

Comment (pkch) [inline pep-0484.txt]: Hmm, I am not sure.

Somehow I would interpret `instance attribute` as referring to the physical object stored on an instance, and of course if there is one, we can't actually refer to the class attribute at all.

File: pep-0484.txt (hunk 1 above)
Comment (pkch) [inline pep-0484.txt]: Oh I thought so too, but I searched the text and found it uses `parametrized`. Now I searched again, and I found it uses both, 2 times each (not counting my own usage) 😆 I guess I should have checked dictionary instead ...

**Comment by pkch** [comment]
I added another clarification: that `TypeVar` cannot contain type variables (based on the discussion in https://github.com/python/mypy/issues/2756#issuecomment-293109162). If needed, I can separate it into a separate PR.

**Comment by pkch** [comment]
The last commit is as discussed [here](https://github.com/python/mypy/issues/3148#issuecomment-292834046)

**Comment by ilevkivskyi** [comment]
@gvanrossum mentioning you just in case (this is not urgent).

File: pep-0484.txt [hunk 3]
```diff
@@ -610,6 +612,24 @@ the runtime class of the objects created by instantiating them doesn't
 record the distinction. This behavior is called "type erasure"; it is
 common practice in languages with generics (e.g. Java, TypeScript).
 
+Using generic classes (parameterized or not) to access attributes will result
+in both type check failure and runtime error. Outside the class definition
```
Comment (ilevkivskyi) [inline pep-0484.txt]: > and runtime error.

Is it really a runtime error for examples below? I just tried it and all worked fine _at runtime_.

File: pep-0484.txt [hunk 4]
```diff
@@ -610,6 +612,24 @@ the runtime class of the objects created by instantiating them doesn't
...
+  Node[int].x = 1  # Error
+  Node[int].x      # Error
+  Node.x = 1       # Error
+  Node.x           # Error
+  type(p).x        # Error
+  p.x              # Ok (evaluates to None)
+  Node[int]().x    # Ok (evaluates to None)
+  p.x = 1          # Ok, but assigning to instance attribute
+
+In addition, ``ClassVar`` cannot be parameterized.
```
Comment (ilevkivskyi) [inline pep-0484.txt]: ``ClassVar`` is not a part of PEP 484.

File: pep-0484.txt [hunk 5]
```diff
@@ -701,11 +721,11 @@ classes without a metaclass conflict.
...
-A type variable may specify an upper bound using ``bound=<type>``.
-This means that an actual type substituted (explicitly or implicitly)
-for the type variable must be a subtype of the boundary type.  A
-common example is the definition of a Comparable type that works well
-enough to catch the most common errors::
+A type variable may specify an upper bound using ``bound=<type>`` (note:
+<type> itself cannot be parametrized by type variables). This means that an
+actual type substituted (explicitly or implicitly) for the type variable must
+be a subtype of the boundary type. A common example is the definition of a
+Comparable type that works well enough to catch the most common errors::
```
Comment (ilevkivskyi) [inline pep-0484.txt]: While yo at it, can you please make this
```
``Comparable``
```
i.e. use backquotes.

File: pep-0484.txt (hunk 3 above)
Comment (pkch) [inline pep-0484.txt]: Ah right. I misunderstood your comment that [it's fixed](https://github.com/python/mypy/issues/2878#issuecomment-282302705) in `typing`. I thought you meant it now causes runtime error, but actually what you meant is that it no longer returns different values depending on how it's accessed.

**Comment by gvanrossum** [comment]
Thanks!

**Comment by ilevkivskyi** [comment]
@pkch Would you like to also make a small PR updating the PEP 526 and saying that things like ``ClassVar[T]`` are prohibited?

**Pull request merged**